src/libstd/sync/mpsc/spsc_queue.rs

```diff
@@ -134,18 +136,25 @@ impl<T, ProducerAddition, ConsumerAddition> Queue<T, ProducerAddition, ConsumerA
 
     unsafe fn alloc(&self) -> *mut Node<T> {
         // First try to see if we can consume the 'first' node for our uses.
-        if *self.producer.first.get() != *self.producer.tail_copy.get() {
-            let ret = *self.producer.first.get();
-            *self.producer.0.first.get() = (*ret).next.load(Ordering::Relaxed);
-            return ret;
-        }
-        // If the above fails, then update our copy of the tail and try
-        // again.
-        *self.producer.0.tail_copy.get() = self.consumer.tail_prev.load(Ordering::Acquire);
-        if *self.producer.first.get() != *self.producer.tail_copy.get() {
-            let ret = *self.producer.first.get();
-            *self.producer.0.first.get() = (*ret).next.load(Ordering::Relaxed);
-            return ret;
+        // SAFEY: Since `self` is a valid reference, accessing its inner values
+        // can be considered safe (without checking for NUL pointers).
+        //
+        // Dereferencing of `ret` are safe too because the pointer comes from
+        // `self` once again.
```

Comment:
This is definitely not true -- we know that accessing the fields of self is valid, but those fields are raw pointer so the validity of the struct itself does not guarantee validity of those pointers. There will likely have to be some investigation done as to whether we're just passing the obligation up to the caller or what here.

Reply:
I read the code more thoroughly and I'm very unsure about it. It seems possible to me this could dereference a `NUL`. It doesn't seems to be a known bug so I'm probably missing something. I changed the comment to mark this.